### PR TITLE
Spike on OCR receipts with Google Cloud Vision API

### PR DESCRIPTION
**Pull request by fgrehm**
As I mentioned over Telegram, I did a quick spike on using [Google's Cloud Vision API](https://cloud.google.com/vision/) to OCR a few receipts and this notebook showcases the results I found. I used different receipts this time and the results look a bit more promising.

Unfortunately it would take me a while to get this right in Python so I just kept it simple and did with just bash. All the files related to it (pdf, payloads and responses) can be grabbed from the Zip below in case someone wants to play with it as well.

[reimbursements-ocr.zip](https://github.com/datasciencebr/serenata-de-amor/files/678854/reimbursements-ocr.zip)

### CODE REVIEW

File: develop/2016-12-30-fgrehm-ocr-receipts-with-google-cloud-vision.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+# As we can see, it takes a while to process just 10 PDFs on a 60Mb connection (5 minutes), if we ever move on with this we should really look into parallelizing it from day 0 and / or sending receipts in batches as it is supported by the API.
+# 
+# ## Document [5631309](https://jarbas.datasciencebr.com/#/document_id/5631309)
+# 
+# A hotel receipt with lots of text in it. There is a lot of stuff on those JSON responses so we just extract the info for the piece that represents the whole text.
+
+# In[3]:
+
+get_ipython().run_cell_magic('bash', '', "jq -r '.responses[].textAnnotations[0].description' /tmp/reimbursements/5631309/response-page-1.json\necho '--------------'; echo\njq -r '.responses[].textAnnotations[0].description' /tmp/reimbursements/5631309/response-page-2.json")
```
Comment (Irio) [inline develop/2016-12-30-fgrehm-ocr-receipts-with-google-cloud-vision.py]: For comparison, using the same method as https://github.com/datasciencebr/serenata-de-amor/commit/fb93f96e334c46f98eea3d4a9db565b8bc6bb45b

```python
import os.path
import subprocess

from PIL import Image
from pytesseract import image_to_string

def pdf_to_image(input_path):
  output_path = input_path.replace('pdf', 'png')
  if not os.path.exists(output_path):
    subprocess.call(['sips', '-s', 'format', 'png',
                     input_path, '--out', output_path])

pdf_to_image('5631309.pdf')
image_to_string(Image.open(input_path), config='-psm 6')
```

gives the following output:

```
Castro Marques Hotels LTDA.
Extrato de Conta
Nome: omvo BILAC 911410 Uh: 1001
Empress: 511311an Ranerva; 1820028
mm can: mass. Iiscal : micrl/CzZ: mom
EnderegomuA 19151115110 mums mum, 4n Apto sol . savanna FunaionéxiozADSIL‘lA
mm HORIZONTE NC 30320-5911 3511st 211113350: 13/03/2015 no t 35
Chagada: 12/c'3/2m: 00:22 Pam’s; 14/03/2915 new:
1
N
CONTA ENCEHHADA Héspeae: PINTO‘ OLAVO BHAC Num Dom «55mm DnslgnagaoﬁLAVO BILAC PINTO
Olin 07mm Documemo on Mamas) Empress Seldo Usuéria
:s/msm m: Iguanas-vacuum mums 71,00 vs,» um:
mm m mm may»: mun noon a“ man ~ awn
Damn/w: 1m - 11,00
momma can. mum “mus "Immun- nma um: ALuzz
Jam am on "cam “1‘59 - 25,17: 1 .
Am cm was 2mm!» :a 4,0: - 12,3:
H/OJI‘IZ? COOI ”VHSNIIII'AWXI‘TI IUWB 259100 381700 NJ!!!
am: warn ms 341m mm In A 1;: 15,170 - 115, cc
:15me as Maw Mam ’Wm H M95 - ma:-
"mm mug mm: mm mm as Inna: - 4m:
(Gan REM ms new?! Jar/LN - 9,09
Aom cm at}; 3:55:12; 2; 4,69 - 9,5:
sacs: 1:; may; {4'91 545: = 1m:
1 as: name by 3,3: - 1mg
1 ”names com 63021119924 921212112; memmw 34,417 nun uwzz
‘ saw mu m 310L513 1x345 :- 14, :0
‘ 3C0”). (UK £1.43 RIOLEVX 5x @530 B 26%;};
:zwsmsa ego: ssuzsrxmmanz 1,0; 123,113 "shunt!
KEY/A C05 “3 BICLEW l x {.06 ~ «351?
13/032365 0001 6903a “kam'tﬂ wanna 192ml} tit/(.00 NNRGEH
5030 as 45mm 5»: 5,50 = 27.37
311m vb maven 3205,50 a 15:55
SizesSrM mt we max/mo zxsmw = 13,051
prom Am- max new saga; a; 1x45, 0: » 49,00
awn G952! a0 mm; can any cs lx§5,&'9 - 45.»
can: mam: a); 3,259 .. no:
1419305163 cm: 69638 X008 SERVICE RESTAUHKRT 50.9%} 19¢,” nut:
Elf-K P1. KLVM MIRCIJA'E PLAZA (6634’ 3X 30190 = 96070
11/03 09:03 xxwxmxxxx. DEPOSIN ANTECIFMEO ~165,€0 ~‘!.IB 1|le
[claim 439133
Resume do Extrato
Héspede: PINTO. OLAVO an Designaoﬁo: OLAVO BHAC PWTO
Descrigéo Valor Total
DEPOSITO AWECIFA 46:. do
FEIGDEAR ;,oa
RES7AUMNTE UJGAN 52a, 00
ROOM same: REST 36¢, 09
R.—-\___U._un_. . K
```

File: develop/2016-12-30-fgrehm-ocr-receipts-with-google-cloud-vision.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+# # Setup
+# 
+# Get an API key from Google Cloud that is able to access the [Vision API](https://cloud.google.com/vision/) and update the snippet below.
+# 
+# **MAKE SURE YOU DON'T COMMIT IT TO VERSION CONTROL**
+
+# In[1]:
+
+get_ipython().run_cell_magic('bash', '', '\nKEY="YOUR_KEY_HERE"\nKEY_PATH="/tmp/cloud-vision.key"\n\nif ! [[ -f $KEY_PATH ]]; then\n  echo "${KEY}" > $KEY_PATH\n  echo "Writing key to ${KEY_PATH}"\nelse\n  echo "Reusing existing key at ${KEY_PATH}"\nfi')
```
Comment (fgrehm) [inline develop/2016-12-30-fgrehm-ocr-receipts-with-google-cloud-vision.py]: Yup, its an alphanum with 39 characters, the only problem is that the notebook is written in bash. Will think about a way of making it work with minimal impact tomorrow.

From your end, you might need to enable cloud vision API for the project associated with the keys you're currently using for google maps. As a side note, I'm willing to share my keys with you and "donate" my US$300 free credit to the project in order to get all receipts OCRed if that's in the plans for the next 2 months before the credit expires 💵

File: develop/2016-12-30-fgrehm-ocr-receipts-with-google-cloud-vision.py (hunk 1 above)
Comment (fgrehm) [inline develop/2016-12-30-fgrehm-ocr-receipts-with-google-cloud-vision.py]: Nice! So I guess that the Cloud Vision API beats tesseract 💪 (at least without any additional tuning of course)

File: develop/2016-12-30-fgrehm-ocr-receipts-with-google-cloud-vision.py (hunk 2 above)
Comment (fgrehm) [inline develop/2016-12-30-fgrehm-ocr-receipts-with-google-cloud-vision.py]: Done on 0fa00a0

**Comment by fgrehm** [comment]
@Irio changes have been made!

**Pull request merged**